staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go

```diff
@@ -177,119 +202,187 @@ func TestGracefulTerminationWithKeepListeningDuringGracefulTerminationDisabled(t
+		resultGot := doer.Do(connReusingClient, func(httptrace.GotConnInfo) {}, "/readyz", time.Second)
+		// wait until we have a non 200 response
+		if resultGot.response != nil && resultGot.response.StatusCode == http.StatusOK {
+			return false, nil
+		}
 
-	// step 4: wait for the HTTP Server listener to have stopped
-	httpServerStoppedListeningCh := s.lifecycleSignals.HTTPServerStoppedListening
-	select {
-	case <-httpServerStoppedListeningCh.Signaled():
-	case <-time.After(5 * time.Second):
-		t.Fatal("Expected the server to signal HTTPServerStoppedListening event")
+		if err := assertResponseStatusCode(resultGot, http.StatusInternalServerError); err != nil {
+			return true, err
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("Expected /readyz to return 500 status code, but got: %v", err)
 	}
 
-	// step 5: the server has stopped listening but we still have a request
+	// before ShutdownDelayDuration elapses new request(s) should be served successfully.
+	beforeShutdownDelayDurationStep.execute(func() {
+		t.Log("Before ShutdownDelayDuration elapses new request(s) should be served")
+		resultGot := doer.Do(connReusingClient, shouldReuseConnection(t), "/echo?message=request-on-an-existing-connection-should-succeed", time.Second)
+		if err := assertResponseStatusCode(resultGot, http.StatusOK); err != nil {
+			t.Errorf("%s", err.Error())
+		}
+		resultGot = doer.Do(newClient(true), shouldUseNewConnection(t), "/echo?message=request-on-a-new-tcp-connection-should-succeed", time.Second)
+		if err := assertResponseStatusCode(resultGot, http.StatusOK); err != nil {
+			t.Errorf("%s", err.Error())
+		}
+	})
+
+	waitForeverUntilSignaled(t, signals.AfterShutdownDelayDuration)
+
+	// preshutdown hook has not completed yet, new incomng request should succeed
+	resultGot = doer.Do(newClient(true), shouldUseNewConnection(t), "/echo?message=request-on-a-new-tcp-connection-should-succeed", time.Second)
+	if err := assertResponseStatusCode(resultGot, http.StatusOK); err != nil {
+		t.Errorf("%s", err.Error())
+	}
+
+	// let the preshutdown hook issue an API call now, and then
+	// let's wait for it to return the result.
+	close(preShutdownHook.blockedCh)
+	preShutdownHookResult := <-preShutdownHook.resultCh
+	if err := assertResponseStatusCode(preShutdownHookResult, http.StatusOK); err != nil {
+		t.Errorf("%s", err.Error())
+	}
+
+	waitForeverUntilSignaled(t, signals.PreShutdownHooksCompleted)
+	// both AfterShutdownDelayDuration and PreShutdownHooksCompleted
+	// have been signaled, we should not be accepting new request
+	waitForeverUntilSignaled(t, signals.NotAcceptingNewRequest)
+	waitForeverUntilSignaled(t, signals.HTTPServerStoppedListening)
+
+	// our original intention was for any incoming request to receive a 503
+	// via the WithWaitGroup filter, but, at this point, any incoming requests
+	// will get a 'connection refused' error since the net/http server has
+	// stopped listening.
```

Comment:
Hmm - is it this PR that is changing this? Or was it always the case?

Reply:
no, this PR is not changing that, it was always the case, before - between the `Server.Shutdown`  <--> in-flight requests drained window, new incoming request(s) end up getting connection refused error.

Reply:
I would add this comment in a separate PR before this one then (and probably not in the test, but rather in the main code).